lib/github-glue.ts

```diff
@@ -108,13 +108,34 @@ export class GitHubGlue {
 
         const url = GitGitGadget.parsePullRequestURL(pullRequestURL);
         const pr = await this.getPRInfo(url[0], url[2]);
-        const ccNow = pr.body.match(/\r?\n\r?\n(cc:.*)/is);
+        const index = pr.body.search(/\r?\n\s*?\r?\n/);
+
+        let found = false;
+        let footerSeparator = "\r\n";
+        if (index > -1) {
+            const regExId = RegExp( `^cc:.*${id[1]}`, "i");
+            const footers = pr.body.slice(index).trimRight().split(/\r?\n/);
+
+            for (let i=footers.length-1; i>=0; i--) {
+                const footer = footers[i];
+                if (regExId.test(footer)) {
+                    found = true;
+                    footerSeparator = "";
+                    break;
+                } else if (footer.match(/^\s*$/)) { // stop at first empty line
+                    break;
+                } else if (footer.match(/^[-A-Za-z]+:/)) { // body has footers
+                    footerSeparator = "";
+                }
+            }
+        }
```

Comment:
I learned that Javascript often performs better, and is more readable, using a map/reduce approach. Something like this:

```js
let found = false, footersSeparator = "";
const footers = pr.body.split(/\r?\n\s*?\r?\n/).pop();
if (footers) try {
    footersSeparator = "\r\n";
    footers.split(/\r?\n/).reverse().forEach((footer) => {
        const match = footer.match(/^([a-z][-a-z0-9]*):\s*(.*)$/i);
        if (!match)
            throw new Error("No footers block");
        found ||= match[0].toLowerCase() === 'cc' &&
            (match[1].trimRight() === id[1] || match[1].search(`<${id[1]>`) >= 0);
    });
} catch {
    found = false;
    footersSeparator = "";
}

if (!found) {
    const user = await this.getGitHubUserInfo(pr.author);
    if (!user.email || user.email !== id[1]) {
        await this.updatePR(url[0], url[2], `${pr.body}${
            footersSeparator}\r\ncc: ${cc}`);
        await this.addPRComment(pullRequestURL, `User \`${
            cc}\` has been added to the cc: list.`);
    }
}
```

Reply:
Old habits die hard - aversion to `try/catch` (in cpp) due to early performance issues.  Changing to use chaining as suggested - it was not used because I was avoiding try/catch.

Updated code will now complain if the last block in the body happens to contain what looks like footer lines.  For example:

```
some text

last block
todo: fix something
```

A test was added for this.

Reply:
btw, your sample code has `footerSeparator` use reversed (for reference when reviewing code).  Also, checking on user.email is now case insensitive after seeing you added to the `cc:` in your own PR. 😀